doc/monads/st.lean

```diff
@@ -0,0 +1,117 @@
+/-!
+# ST monad
+
+In the previous sections, [StateM](states.lean.md) and [ReaderM](readers.lean.md), you saw how
+`ReaderM` provides readonly state and `StateM` provides updatable state. There is another monad
+called `ST` which stands for "State Transfomer" which is a _stateful computation_: the computation
+is seen as transforming one state into another, where the new state will actually be constructed by
+modifying the old state in place.
+
+This is based on the work of John Launchbury and Simon L Peyton Jones in
+[Lazy Functional State Threads](https://www.microsoft.com/en-us/research/wp-content/uploads/1994/06/lazy-functional-state-threads.pdf).
+
+A state transfomer can be used as a first-class value which means it can be passed to a funciton,
+returned as a result and stored in a data structure, which means you can store it in a `ReaderM`
+context to make part of that context updatable.
+
+## Motivating example: A simple Job Server
+
+This simple server has a `ReaderT ServerContext` that contains a bunch of useful state needed by our
+server function, but it also has a reference to an updateable `List Nat` which is typed as `IO.Ref
+(List Nat)`. `IO.Ref` is an alias for `ST.Ref` which is a mutable reference to an object.
+
+In this example the server will use multi-threading to process all the items in the `args` list in
+parallel, and each `asyncProcess` will create a new thread and use the thread safe `ST.Ref.modify`
+function to update the shared state, which is just a listed of `processed` jobs in our case.
+
+-/
+structure ServerContext where
+   hIn : IO.FS.Stream
+   hOut : IO.FS.Stream
+   args : List Nat
+   processed : IO.Ref (List Nat)
+
+def asyncProcess (i: Nat) : ReaderT ServerContext IO (Task (Except IO.Error Unit)) := do
+  let jobs ← read
+  IO.asTask do
+    IO.sleep 1
+    jobs.processed.modify (λ s => i :: s)
+
+def processTasks (args: List Nat) : ReaderT ServerContext IO Unit := do
+  let tasks ← args.mapM fun s => asyncProcess s
+  for task in tasks do
+    ofExcept task.get
+
+def server (args : List Nat) : IO (List Nat) := do
+  let hIn ← IO.getStdin
+  let hOut ← IO.getStdout
+  let processedRef : IO.Ref (List Nat) ← IO.mkRef []
+  let context : ServerContext := { hIn, hOut, args, processed := processedRef}
+  processTasks args |>.run context
+  context.processed.get
```

Comment:
```suggestion
def server (args : Array Nat) : IO (Array Nat) := do
  -- make an array for the completed results
  let processed : IO.Ref (Array Nat) ← IO.mkRef #[]
  -- each task will run in parallel
  let tasks ← args.mapM fun i => do
    IO.asTask do
      IO.sleep 1 -- do some work
      processed.modify (·.push i)
  -- block on the completion of each task
  for task in tasks do
    ofExcept task.get
  -- now we can read the result
  processed.get
```
The example is overly complicated. It should be reduced to focus on the reference operations.

Reply:
Actually, I don't think so.  The reader just learned about ReaderT and StateT and is now seeing that an updatable ST.Ref can be used in a ReaderT context whereas your version would be great for a chapter purely focused on "references" and doesn't care at all about monads.  Plus your version could also be in a section that talks about the limitations of `let mut` in a mutithreaded app and how an IO.Ref can overcome that limitation.